src/ht.c

```diff
@@ -146,162 +149,198 @@ SDB_API void ht_free(SdbHt* ht) {
+	HtKv *kv_dst = reserve_kv (ht, kv->key, kv->key_len, update);
+	if (!kv_dst) {
 		return false;
 	}
-	HtKv* kv = calloc (1, sizeof (HtKv));
-	if (kv) {
-		kv->key = dupkey (ht, key);
-		kv->value = dupval (ht, value);
-		kv->key_len = calcsize_key (ht, kv->key);
-		kv->value_len = calcsize_val (ht, kv->value);
-
-		if (!internal_ht_insert_kv (ht, kv, update)) {
-			freefn (ht, kv);
-			return false;
-		}
-		return true;
-	}
-	return false;
+
+	memcpy (kv_dst, kv, ht->elem_size);
+	check_growing (ht);
+	return true;
 }
 
-SDB_API bool ht_insert_kv(SdbHt *ht, HtKv *kv, bool update) {
-	return internal_ht_insert_kv (ht, kv, update);
+static bool insert_update(SdbHt *ht, const char *key, void *value, bool update) {
+	ut32 key_len = calcsize_key (ht, key);
+	HtKv* kv_dst = reserve_kv (ht, key, key_len, update);
+	if (!kv_dst) {
+		return false;
+	}
+
+	kv_dst->key = dupkey (ht, key);
+	kv_dst->key_len = key_len;
+	kv_dst->value = dupval (ht, value);
+	kv_dst->value_len = calcsize_val (ht, value);
+	check_growing (ht);
+	return true;
 }
+
 // Inserts the key value pair key, value into the hashtable.
 // Doesn't allow for "update" of the value.
 SDB_API bool ht_insert(SdbHt* ht, const char* key, void* value) {
-	return internal_ht_insert (ht, false, key, value);
+	return insert_update (ht, key, value, false);
 }
 
 // Inserts the key value pair key, value into the hashtable.
 // Does allow for "update" of the value.
 SDB_API bool ht_update(SdbHt* ht, const char* key, void* value) {
-	return internal_ht_insert (ht, true, key, value);
+	return insert_update (ht, key, value, true);
 }
 
 // Returns the corresponding SdbKv entry from the key.
 // If `found` is not NULL, it will be set to true if the entry was found, false
 // otherwise.
 SDB_API HtKv* ht_find_kv(SdbHt* ht, const char* key, bool* found) {
-	if (!ht) {
-		return NULL;
-	}
 	ut32 bucket = bucketfn (ht, key);
-	SdbListIter* iter;
-	HtKv* kv;
+
+	if (found) {
+		*found = false;
+	}
+
+	HtBucket *bt = &ht->table[bucket];
 	ut32 key_len = calcsize_key (ht, key);
+	HtKv *kv;
+	ut32 j;
 
-	ls_foreach (ht->table[bucket], iter, kv) {
+	BUCKET_FOREACH (ht, bt, j, kv) {
 		if (is_kv_equal (ht, key, key_len, kv)) {
 			if (found) {
 				*found = true;
 			}
 			return kv;
 		}
 	}
-	if (found) {
-		*found = false;
-	}
 	return NULL;
 }
 
 // Looks up the corresponding value from the key.
 // If `found` is not NULL, it will be set to true if the entry was found, false
 // otherwise.
 SDB_API void* ht_find(SdbHt* ht, const char* key, bool* found) {
-	bool _found = false;
-	if (!found) {
-		found = &_found;
-	}
-	HtKv* kv = ht_find_kv (ht, key, found);
-	return (kv && *found)? kv->value : NULL;
+	HtKv *res = ht_find_kv (ht, key, found);
+	return res ? res->value : NULL;
 }
 
 // Deletes a entry from the hash table from the key, if the pair exists.
 SDB_API bool ht_delete(SdbHt* ht, const char* key) {
-	return ht_delete_internal (ht, key, NULL);
+	ut32 bucket = bucketfn (ht, key);
```

Comment:
no need to hold this in a variable, the return value is only used once and we should trust the return value